cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/cucumber_expression.rb

```diff
@@ -51,7 +52,16 @@ def process_optional(expression)
         # Create non-capturing, optional capture groups from parenthesis
         expression.gsub(OPTIONAL_REGEXP) do
           g2 = $2
-          check_no_parameter_type(g2, PARAMETER_TYPES_CANNOT_BE_OPTIONAL)
+          # When using Parameter Types, the () characters are used to represent an optional
+          # item such as (a ) which would be equivalent to (?:a )? in regex
+          #
+          # You cannot have optional Parameter Types i.e. ({int}) as this causes
+          # problems during the conversion phase to regex. So we check for that here
+          #
+          # One exclusion to this rule is if you actually want the brackets i.e. you
+          # want to capture (3) then we still permit this as an individual rule
+          # See: https://github.com/cucumber/cucumber-ruby/issues/1337 for more info
+          check_no_parameter_type(g2, PARAMETER_TYPES_CANNOT_BE_OPTIONAL) unless brackets_wanted?(expression)
```

Comment:
Not sure if this is the right way to write this in ruby but why write `unless $1 == DOUBLE_ESCAPE`? I can't come up with a test case that would make this fail.

Reply:
Is that a strong enough test? And I've done something similar just by using another private method.

Don't we also need to check there is a parameter being captured in the middle? Or is that kind of redundant?

Reply:
> Don't we also need to check there is a parameter being captured in the middle? Or is that kind of redundant?

I think so. We are escaping the brackets. So we don't have to check if there is a parameter type in the brackets because as far as Cucumber is concerned there are none (they are escaped).

So for example test case should also pass `some \(prefix {int} infix {int} suffix) thing` and this too `some \(prefix {int} infix {int} infix {int} infix {int} infix {int} suffix) thing`.

Reply:
Maybe it'd be less confusing to write:
```
if $1 == DOUBLE_ESCAPE:
   return "\\(#{g2}\\)"
else 
  check_no_parameter_type(g2, PARAMETER_TYPES_CANNOT_BE_OPTIONAL)
  return "(?:#{g2})?"

```

Reply:
:point_up: looks good. Let me take another attack later; I'll also write some more specs.